external-crates/move/crates/move-compiler/src/typing/macro_expand.rs

```diff
@@ -0,0 +1,955 @@
+            let recolor = &mut Recolor::new(
+                next_color,
+                /* return already labeled */ None,
+                recolor_use_funs,
+            );
+            recolor_exp(recolor, &mut arg);
+            context.core.set_max_variable_color(recolor.max_color());
+
+            // mark the arg as coming from an argument substitution for recursive checks
+            match &mut arg.value {
+                N::Exp_::Block(block) => {
+                    block.from_macro_argument = Some(N::MacroArgument::Substituted(*eloc))
+                }
+                N::Exp_::UnresolvedError => (),
+                _ => unreachable!("ICE all macro args should have been made blocks in naming"),
+            };
+
+            *e_ = N::Exp_::Annotate(Box::new(arg), expected_ty);
+        }
+        N::Exp_::VarCall(sp!(_, v_), _) if context.by_name_args.contains_key(v_) => {
+            context.mark_used(v_);
+            let (arg, _expected_ty) = context.by_name_args.get(v_).unwrap();
+            context.core.env.add_diag(diag!(
+                TypeSafety::CannotExpandMacro,
+                (*eloc, "Cannot call non-lambda argument"),
+                (arg.loc, "Expected a lambda argument")
+            ));
+            *e_ = N::Exp_::UnresolvedError;
+        }
+
+        ///////
+        // Other var cases
+        ///////
+        N::Exp_::Var(sp!(_, v_)) => {
+            let is_unbound_param = context
+                .all_params
+                .get(v_)
+                .is_some_and(|info| info.argument.is_none());
+            if is_unbound_param {
+                assert!(!context.lambdas.contains_key(v_));
+                assert!(!context.by_name_args.contains_key(v_));
+                assert!(
+                    context.core.env.has_errors(),
+                    "ICE unbound param should have already resulted in an error"
+                );
+                *e_ = N::Exp_::UnresolvedError;
+            }
+        }
+        N::Exp_::VarCall(sp!(_, v_), sp!(_, es)) => {
+            exps(context, es);
+            let is_unbound_param = context
+                .all_params
+                .get(v_)
+                .is_some_and(|info| info.argument.is_none());
+            if is_unbound_param {
+                assert!(!context.lambdas.contains_key(v_));
+                assert!(!context.by_name_args.contains_key(v_));
+                assert!(
+                    context.core.env.has_errors(),
+                    "ICE unbound param should have already resulted in an error"
+                );
+                *e_ = N::Exp_::UnresolvedError;
+            }
+        }
+    }
+}
+
+fn builtin_function(context: &mut Context, sp!(_, bf_): &mut N::BuiltinFunction) {
+    match bf_ {
+        N::BuiltinFunction_::Freeze(ty_opt) => {
+            if let Some(ty) = ty_opt {
+                type_(context, ty)
+            }
+        }
+        N::BuiltinFunction_::Assert(_) => todo!(),
```

Comment:
Does this need to get done before landing?

Reply:
I think I put this bomb here because I was thinking about expanding these during macro expansion, but we should punt it for now. Good catch 😅